File: 1169. Invalid Transactions.py
class Solution:
    class Node:
        def __init__(self, s: str):
            self.name, t, v, self.dst = s.split(",");
            self.ori = s
            self.t = int(t)
            self.v = int(v)
            
        
    def invalidTransactions(self, ts: List[str]) -> List[str]:
        n = len(ts)
        a = sorted([self.Node(s) for s in ts], key = lambda node: node.t)
        
        q = collections.deque()
        ans = set()
        h = collections.defaultdict(set)
        for i, x in enumerate(a):
            while len(q) and a[q[0]].t < x.t-60:
                h[a[q[0]].name].remove(q[0])
                q.popleft()
            if x.v > 1000:
                ans.add(i)
            for p in h[x.name]:
                if a[p].dst != x.dst:
                    ans.add(p)
                    ans.add(i)
            q.append(i)
            h[x.name].add(i);
        
        return [a[x].ori for x in ans]
        
        

File: 1169. Invalid Transactions.cpp
class Solution {
public:
    
    struct Node {
        string name, dst;
        int time, val;
        Node(string name_, int time_, int val_, string dst_):
            name(std::move(name_)), time(time_), val(val_), dst(std::move(dst_)) {}
        
        string to_string() {
            return name + "," + std::to_string(time) + "," + std::to_string(val) + "," + dst;
        }
    };
    
    vector<string> invalidTransactions(vector<string>& ss) {
        vector<Node> a;
        unordered_set<int> ans;
        queue<int> q;
        unordered_map<string, unordered_set<int>> h;
        for (const auto& s : ss) {
            istringstream sin(s);
            string name, time, val, dst;
            getline(sin, name, ',');
            getline(sin, time, ',');
            getline(sin, val, ',');
            getline(sin, dst, ',');
            a.push_back(Node(name, stoi(time), stoi(val), dst));
        }
        
        sort(a.begin(), a.end(), [](Node& a, Node& b) {
            return a.time < b.time;
        });
        
        for (int id=0; id<a.size(); ++id) {
         
            while (q.size() && a[q.front()].time < a[id].time-60) {
                h[a[q.front()].name].erase(q.front());
                q.pop();
            }
            
            if (a[id].val > 1000) {
                ans.insert(id);
            }
            
            for (auto i : h[a[id].name])
                if (a[id].dst != a[i].dst) {
                    ans.insert(id);
                    ans.insert(i);
                }
            q.push(id);
            h[a[id].name].insert(id);
        }
        
        vector<string> ret;
        for (auto x : ans) ret.push_back(a[x].to_string());
        return ret;
    }
};